File: interfaceTest/ssros_ams/src/spin_action_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "test_interfaces/action/spin.hpp"

#include <memory>
#include <string>
#include <iostream>

class SpinActionClient : public rclcpp::Node
{
public:
  using Spin = test_interfaces::action::Spin;
  using GoalHandleSpin = rclcpp_action::ClientGoalHandle<Spin>;

  explicit SpinActionClient(const rclcpp::NodeOptions & node_options = rclcpp::NodeOptions())
  : Node("spin_action_client", node_options)
  {
    this->client_ptr_ = rclcpp_action::create_client<Spin>(
      this,
      "spin");

    this->timer_ = this->create_wall_timer(
      std::chrono::milliseconds(500),
      std::bind(&SpinActionClient::send_goal, this));
  }

  void send_goal()
  {
    using namespace std::placeholders;

    this->timer_->cancel();

    if (!this->client_ptr_->wait_for_action_server()) {
      RCLCPP_ERROR(this->get_logger(), "Action server not available after waiting");
      rclcpp::shutdown();
    }

    auto goal_msg = Spin::Goal();
    goal_msg.target_yaw = 10.0;

    RCLCPP_INFO(this->get_logger(), "Sending goal");

    auto send_goal_options = rclcpp_action::Client<Spin>::SendGoalOptions();
    send_goal_options.goal_response_callback =
      std::bind(&SpinActionClient::goal_response_callback, this, _1);
    send_goal_options.feedback_callback =
      std::bind(&SpinActionClient::feedback_callback, this, _1, _2);
    send_goal_options.result_callback =
      std::bind(&SpinActionClient::result_callback, this, _1);
    this->client_ptr_->async_send_goal(goal_msg, send_goal_options);
  }

private:
  rclcpp_action::Client<Spin>::SharedPtr client_ptr_;
  rclcpp::TimerBase::SharedPtr timer_;

  void goal_response_callback(const GoalHandleSpin::SharedPtr & goal_handle)
  {
    if (!goal_handle) {
      RCLCPP_ERROR(this->get_logger(), "Goal was rejected by server");
    } else {
      RCLCPP_INFO(this->get_logger(), "Goal accepted by server, waiting for result");
    }
  }

  void feedback_callback(
    GoalHandleSpin::SharedPtr,
    const std::shared_ptr<const Spin::Feedback> feedback)
  {
    RCLCPP_INFO(
      this->get_logger(),
      "Feedback received: %f",
      feedback->angular_distance_traveled);
  }

  void result_callback(const GoalHandleSpin::WrappedResult & result)
  {
    switch (result.code) {
      case rclcpp_action::ResultCode::SUCCEEDED:
        break;
      case rclcpp_action::ResultCode::ABORTED:
        RCLCPP_ERROR(this->get_logger(), "Goal was aborted");
        return;
      case rclcpp_action::ResultCode::CANCELED:
        RCLCPP_ERROR(this->get_logger(), "Goal was canceled");
        return;
      default:
        RCLCPP_ERROR(this->get_logger(), "Unknown result code");
        return;
    }
    RCLCPP_INFO(this->get_logger(), "Result received: %d", result.result->error_code);
    rclcpp::shutdown();
  }
};

int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    auto action_client = std::make_shared<SpinActionClient>();
    rclcpp::spin(action_client);
    rclcpp::shutdown();
    return 0;
}

File: interfaceTest/ssros_ams_py/ssros_ams_py/spin_action_client.py
import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node
from test_interfaces.action import Spin

class SpinActionClient(Node):
    def __init__(self):
        super().__init__('spin_action_client')
        self._action_client = ActionClient(self, Spin, 'spin')

    def send_goal(self, target_yaw):
        goal_msg = Spin.Goal()
        goal_msg.target_yaw = target_yaw

        self._action_client.wait_for_server()

        self._send_goal_future = self._action_client.send_goal_async(
            goal_msg,
            feedback_callback=self.feedback_callback)

        self._send_goal_future.add_done_callback(self.goal_response_callback)

    def goal_response_callback(self, future):
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().info('Goal rejected :(')
            return

        self.get_logger().info('Goal accepted :)')

        self._get_result_future = goal_handle.get_result_async()
        self._get_result_future.add_done_callback(self.get_result_callback)

    def get_result_callback(self, future):
        result = future.result().result
        self.get_logger().info('Result: {0}'.format(result.error_code))
        rclpy.shutdown()

    def feedback_callback(self, feedback_msg):
        feedback = feedback_msg.feedback
        self.get_logger().info('Received feedback: {0}'.format(feedback.angular_distance_traveled))

def main(args=None):
    rclpy.init(args=args)
    action_client = SpinActionClient()
    action_client.send_goal(10.0)
    rclpy.spin(action_client)

if __name__ == '__main__':
    main()
